--- google-perftools-1.7/Makefile.am.orig	2011-02-24 15:13:43.022099302 -0800
+++ google-perftools-1.7/Makefile.am	2011-02-24 15:14:03.412082438 -0800
@@ -729,7 +729,7 @@
 debugallocation_test_SOURCES = src/tests/debugallocation_test.cc
 debugallocation_test_CXXFLAGS = $(PTHREAD_CFLAGS) $(AM_CXXFLAGS)
 debugallocation_test_LDFLAGS = $(PTHREAD_CFLAGS) $(TCMALLOC_FLAGS)
-debugallocation_test_LDADD = libtcmalloc_debug.la $(PTHREAD_LIBS)
+debugallocation_test_LDADD = libtcmalloc_minimal_debug.la $(PTHREAD_LIBS)
 
 endif WITH_DEBUGALLOC
 
--- google-perftools-1.7/Makefile.in.orig	2011-02-24 15:13:20.412098139 -0800
+++ google-perftools-1.7/Makefile.in	2011-02-24 15:31:00.032153239 -0800
@@ -790,7 +790,6 @@
 @WITH_DEBUGALLOC_TRUE@am_debugallocation_test_OBJECTS = debugallocation_test-debugallocation_test.$(OBJEXT)
 debugallocation_test_OBJECTS = $(am_debugallocation_test_OBJECTS)
 @WITH_DEBUGALLOC_TRUE@debugallocation_test_DEPENDENCIES =  \
-@WITH_DEBUGALLOC_TRUE@	libtcmalloc_debug.la \
 @WITH_DEBUGALLOC_TRUE@	$(am__DEPENDENCIES_1)
 am__debugallocation_test_sh_SOURCES_DIST =  \
 	src/tests/debugallocation_test.sh
@@ -2135,7 +2134,7 @@
 @WITH_DEBUGALLOC_TRUE@debugallocation_test_SOURCES = src/tests/debugallocation_test.cc
 @WITH_DEBUGALLOC_TRUE@debugallocation_test_CXXFLAGS = $(PTHREAD_CFLAGS) $(AM_CXXFLAGS)
 @WITH_DEBUGALLOC_TRUE@debugallocation_test_LDFLAGS = $(PTHREAD_CFLAGS) $(TCMALLOC_FLAGS)
-@WITH_DEBUGALLOC_TRUE@debugallocation_test_LDADD = libtcmalloc_debug.la $(PTHREAD_LIBS)
+@WITH_DEBUGALLOC_TRUE@debugallocation_test_LDADD = libtcmalloc_minimal_debug.la $(PTHREAD_LIBS)
 
 ### ------- tcmalloc (thread-caching malloc + heap profiler + heap checker)
 
