--- opendkim-2.7.0/opendkim/opendkim-db.c.orig	2012-10-29 13:39:06.893682984 -0700
+++ opendkim-2.7.0/opendkim/opendkim-db.c	2012-10-29 13:41:06.553680091 -0700
@@ -2860,19 +2860,22 @@
 		}
 
 		lderr = dkimf_db_open_ldap(&ld, ldap, err);
-		if (lderr != LDAP_SUCCESS &&
-		    (flags & DKIMF_DB_FLAG_SOFTSTART) == 0)
-		{
-			if (err != NULL)
-				*err = ldap_err2string(lderr);
-			free(ldap);
-			free(p);
-			free(new);
-			return -1;
-		}
-		else
+		if (lderr != LDAP_SUCCESS)
 		{
 			ld = NULL;
+			if ((flags & DKIMF_DB_FLAG_SOFTSTART) == 0)
+			{
+				if (err != NULL)
+					*err = ldap_err2string(lderr);
+				free(ldap);
+				free(p);
+				free(new);
+				return -1;
+			}
+			else
+			{
+				ld = NULL;
+			}
 		}
 
 		pthread_mutex_init(&ldap->ldap_lock, NULL);
