--- opendkim-2.6.0/opendkim/opendkim-db.c.orig	2012-05-08 16:16:12.638313716 -0700
+++ opendkim-2.6.0/opendkim/opendkim-db.c	2012-05-08 16:16:29.828339074 -0700
@@ -4756,6 +4756,13 @@
 			return dkimf_db_get(db, buf, buflen, req, reqnum,
 			                    exists);
 		}
+		else if (status == LDAP_TIMEOUT)
+		{
+			ldap_unbind_ext(ld, NULL, NULL);
+			db->db_handle = NULL;
+			return dkimf_db_get(db, buf, buflen, req, reqnum,
+			                    exists);
+		}
 		else if (status != LDAP_SUCCESS)
 		{
 			db->db_status = status;
