--- opendkim-2.5.2/configure.orig	2012-04-26 15:24:40.978476524 -0700
+++ opendkim-2.5.2/configure	2012-04-26 15:27:25.818476301 -0700
@@ -18531,6 +18531,14 @@
 			LIBXML2_CPPFLAGS="-I$xml2path/include"
 			LIBXML2_LDFLAGS="-I$xml2path/lib"
 			LIBXML2_LIBS="-lxml2"
+		elif test -f $xml2path/include/libxml2/libxml/parser.h
+		then
+			{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $xml2path" >&5
+$as_echo "$xml2path" >&6; }
+			xml2_found="yes"
+			LIBXML2_CPPFLAGS="-I$xml2path/include/libxml2"
+			LIBXML2_LDFLAGS="-I$xml2path/lib"
+			LIBXML2_LIBS="-lxml2"
 		else
 			as_fn_error $? "not found at $xml2path" "$LINENO" 5
 		fi
