--- amavisd-new-2.6.2/amavisd.orig	2009-01-28 13:12:19.000000000 -0800
+++ amavisd-new-2.6.2/amavisd	2009-01-28 13:42:07.000000000 -0800
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -T
+#!/usr/bin/perl
 
 #------------------------------------------------------------------------------
 # This is amavisd-new.
@@ -383,6 +383,7 @@
     [qw(
       $helpers_home $dspam $sa_spawned
       $sa_local_tests_only $sa_auto_whitelist $sa_timeout $sa_debug
+	$sa_rules_file $sa_site_rules_file $sa_userprefs_file
     )],
     'platform' => [qw(
       $can_truncate $unicode_aware $eol $my_pid
@@ -844,7 +845,7 @@
   $spam_quarantine_to       = 'spam-quarantine';
   $bad_header_quarantine_to = 'bad-header-quarantine';
   $clean_quarantine_to      = 'clean-quarantine';
-  $archive_quarantine_to    = 'archive-quarantine';
+  $archive_quarantine_to    = undef;
 
   # similar to $spam_quarantine_to, but the lookup key is the sender address
   $spam_quarantine_bysender_to = undef;  # dflt: no by-sender spam quarantine
@@ -1689,9 +1690,9 @@
     elsif ($errn)        { $msg = "is inaccessible: $!" }
     elsif (-d _)         { $msg = "is a directory" }
     elsif (!-f _)        { $msg = "is not a regular file" }
-    elsif ($> && -o _)   { $msg = "should not be owned by EUID $>"}
+#    elsif ($> && -o _)   { $msg = "should not be owned by EUID $>"}
     elsif ($> && -w _)   { $msg = "is writable by EUID $>, EGID $)" }
-    elsif ($owner_uid)   { $msg = "should be owned by root (uid 0) "}
+#    elsif ($owner_uid)   { $msg = "should be owned by root (uid 0) "}
     if (defined $msg)    { die "Config file \"$config_file\" $msg," }
     $read_config_files_depth++;  push(@actual_config_files, $config_file);
     if ($read_config_files_depth >= 100) {
@@ -8921,8 +8922,8 @@
           rename($config_file.'.moved', $config_file);
           $m = 'is writable (confirmed)';
         }
-        push(@msg, "Directory of a config file \"$config_file\" $m, ".
-                   "UID $<, EUID $>, EGID $)" );
+        #push(@msg, "Directory of a config file \"$config_file\" $m, ".
+        #           "UID $<, EUID $>, EGID $)" );
       }
       last  if @msg;
     }
@@ -12368,6 +12369,14 @@
       push(@q_tuples, [$qar_method,$q,'Arch'])  if defined $q && $q ne '';
     }
   }  # endfor per_recip_data
+  if (defined($qar_method) && $qar_method ne '') {  # archiving quarantine on sender
+    my($sender) = $msginfo->sender;
+    if ($sender ne '') {
+      my($q) = lookup(0,$sender,@{ca('archive_quarantine_to_maps')});
+      $q = $sender  if $q ne '' && $qar_method =~ /^bsmtp:/i;  # orig.recip
+      push(@q_tuples, [$qar_method,$q,'Arch'])  if defined $q && $q ne '';
+    }
+  }
 
   if ($ccat == CC_SPAM) {
     my($sqbsm) = ca('spam_quarantine_bysender_to_maps');
@@ -21037,6 +21046,9 @@
     local_tests_only  => $sa_local_tests_only,
     home_dir_for_helpers => $helpers_home,
     stop_at_threshold => 0,
+    rules_filename => $sa_rules_file,
+    site_rules_filename => $sa_site_rules_file,
+    userprefs_filename => $sa_userprefs_file,
     need_tags         => 'TIMING,LANGUAGES,RELAYCOUNTRY,ASN,ASNCIDR',
 #   LOCAL_STATE_DIR   => '/var/lib',
 #   PREFIX            => '/usr/local',
--- amavisd-new-2.6.2/LDAP.ldif.orig	2009-01-28 14:20:45.000000000 -0800
+++ amavisd-new-2.6.2/LDAP.ldif	2009-01-28 14:20:53.000000000 -0800
@@ -31,7 +31,7 @@
 
 dn: cn=amavisd,cn=schema,cn=config
 objectClass: olcSchemaConfig
-cn: cn=amavisd
+cn: amavisd
 olcAttributeTypes: {0}( 1.3.6.1.4.1.15312.2.2.1.1 NAME 'amavisVirusLover' DESC
   'Virus Lover' EQUALITY booleanMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.7 SING
  LE-VALUE )
