--- amavisd-new-2.4.1/amavisd.fcs	2006-06-23 10:34:28.000000000 -0700
+++ amavisd-new-2.4.1/amavisd	2006-06-23 10:37:17.000000000 -0700
@@ -1,4 +1,4 @@
-#!/usr/bin/perl -T
+#!/usr/bin/perl
 
 #------------------------------------------------------------------------------
 # This is amavisd-new.
@@ -258,6 +258,7 @@
       $file
     )],
     'sa' => [qw(
+	  $sa_rules_file $sa_site_rules_file $sa_userprefs_file
       $helpers_home $dspam
       $sa_local_tests_only $sa_auto_whitelist $sa_timeout $sa_debug
     )],
@@ -1196,7 +1197,6 @@
     elsif ($errn)      { $msg = "is inaccessible: $!" }
     elsif (-d _)       { $msg = "is a directory" }
     elsif (!-f _)      { $msg = "is not a regular file" }
-    elsif ($> && -o _) { $msg = "is owned by EUID $>, should be owned by root"}
     elsif ($> && -w _) { $msg = "is writable by EUID $>, EGID $)" }
     if (defined $msg)  { die "Config file \"$config_file\" $msg," }
     $! = 0;
@@ -6715,8 +6715,8 @@
           rename($config_file.'.moved', $config_file);  # try to rename back
           $m = 'is writable (confirmed)';
         }
-        push(@msg, "Directory of a config file \"$config_file\" $m, ".
-                   "UID $<, EUID $>, EGID $)" );
+#        push(@msg, "Directory of a config file \"$config_file\" $m, ".
+#                   "UID $<, EUID $>, EGID $)" );
       }
       last  if @msg;
     }
@@ -14760,6 +14760,9 @@
     local_tests_only  => $sa_local_tests_only,
     home_dir_for_helpers => $helpers_home,
     stop_at_threshold => 0,
+	rules_filename => $sa_rules_file,
+	site_rules_filename => $sa_site_rules_file,
+	userprefs_filename => $sa_userprefs_file,
 #   DEF_RULES_DIR     => '/usr/local/share/spamassassin',
 #   LOCAL_RULES_DIR   => '/etc/mail/spamassassin',
 #see man Mail::SpamAssassin for other options
