--- postfix-2.11-20130927/src/util/dict_lmdb.c.orig	2013-10-15 08:37:46.005692231 -0700
+++ postfix-2.11-20130927/src/util/dict_lmdb.c	2013-10-15 08:38:55.109689929 -0700
@@ -921,7 +921,7 @@
     dict_lmdb->dict.delete = dict_lmdb_delete;
     dict_lmdb->dict.sequence = dict_lmdb_sequence;
     dict_lmdb->dict.close = dict_lmdb_close;
-    if ((dict_lmdb->dict.stat_fd = open(mdb_path, O_RDONLY)) < 0)
+    if ((dict_lmdb->dict.stat_fd = open(mdb_path, O_RDWR)) < 0)
 	msg_fatal("dict_lmdb_open: %s: %m", mdb_path);
     if (fstat(dict_lmdb->dict.stat_fd, &st) < 0)
 	msg_fatal("dict_lmdb_open: fstat: %m");
