--- postfix-2.11-20130818.2z/makedefs.orig	2013-08-19 09:52:02.549886882 -0700
+++ postfix-2.11-20130818.2z/makedefs	2013-08-19 09:53:07.189887982 -0700
@@ -286,6 +286,7 @@
 		case "$CCARGS" in
 		 *-DNO_DB*) ;;
 		 *-DHAS_DB*) ;;
+		 *-DHAS_LMDB*) ;;
 		 *) if [ -f /usr/include/db.h ]
 		    then
 			: we are all set
@@ -359,6 +360,7 @@
 		case "$CCARGS" in
 		 *-DNO_DB*) ;;
 		 *-DHAS_DB*) ;;
+		 *-DHAS_LMDB*) ;;
 		 *) if [ -f /usr/include/db.h ]
 		    then
 			: we are all set
