--- postfix-2.8.5/src/global/dict_ldap.c.orig	2011-10-28 12:45:37.559023234 -0700
+++ postfix-2.8.5/src/global/dict_ldap.c	2011-10-28 14:33:43.689021170 -0700
@@ -498,6 +498,7 @@
 static int dict_ldap_result(LDAP *ld, int msgid, int timeout, LDAPMessage **res)
 {
     struct timeval mytimeval;
+    int err;
 
     mytimeval.tv_sec = timeout;
     mytimeval.tv_usec = 0;
@@ -506,9 +507,12 @@
     if (ldap_result(ld, msgid, GET_ALL, &mytimeval, res) == -1)
 	return (dict_ldap_get_errno(ld));
 
-    if (dict_ldap_get_errno(ld) == LDAP_TIMEOUT) {
-	(void) dict_ldap_abandon(ld, msgid);
-	return (dict_ldap_set_errno(ld, LDAP_TIMEOUT));
+    if ((err = dict_ldap_get_errno(ld)) != LDAP_SUCCESS) {
+      if (err == LDAP_TIMEOUT) {
+        (void) dict_ldap_abandon(ld, msgid);
+        return (dict_ldap_set_errno(ld, LDAP_TIMEOUT));
+      }
+      return err;
     }
     return LDAP_SUCCESS;
 }
@@ -552,6 +556,7 @@
 static int dict_ldap_bind_st(DICT_LDAP *dict_ldap)
 {
     int     rc;
+    int     err = LDAP_SUCCESS;
     int     msgid;
     LDAPMessage *res;
     struct berval cred;
@@ -567,7 +572,8 @@
 	return (rc);
 
 #define FREE_RESULT 1
-    return (ldap_parse_sasl_bind_result(dict_ldap->ld, res, 0, FREE_RESULT));
+    rc = ldap_parse_result(ld, res, &err, 0, 0, 0, 0, FREE_RESULT);
+    return (rc == LDAP_SUCCESS ? err : rc);
 }
 
 /* search_st - Synchronous search with timeout */
