--- postfix-2.10-20120422/makedefs.orig	2012-04-24 13:39:46.854977395 -0700
+++ postfix-2.10-20120422/makedefs	2012-04-24 13:42:23.159145449 -0700
@@ -277,7 +277,10 @@
     Linux.2*)	SYSTYPE=LINUX2
 		case "$CCARGS" in
 		 *-DNO_DB*) ;;
-		 *) if [ -f /usr/include/db.h ]
+		 *) if [ -f /opt/zimbra/bdb/include/db.h ]
+                    then
+                        CCARGS="$CCARGS -I/opt/zimbra/bdb/include"
+                    elif [ -f /usr/include/db.h ]
 		    then
 			: we are all set
 		    elif [ -f /usr/include/db/db.h ]
@@ -349,7 +352,10 @@
     Linux.3*)	SYSTYPE=LINUX3
 		case "$CCARGS" in
 		 *-DNO_DB*) ;;
-		 *) if [ -f /usr/include/db.h ]
+		 *) if [ -f /opt/zimbra/bdb/include/db.h ]
+                    then
+                        CCARGS="$CCARGS -I/opt/zimbra/bdb/include"
+                    elif [ -f /usr/include/db.h ]
 		    then
 			: we are all set
 		    elif [ -f /usr/include/db/db.h ]
@@ -367,7 +373,7 @@
 		esac
 		for name in nsl resolv
 		do
-		    for lib in /usr/lib64 /lib64 /usr/lib /lib
+		    for lib in /usr/lib64 /lib64 /usr/lib /lib /usr/lib/x86_64-linux-gnu /lib/x86_64-linux-gnu
 		    do
 			test -e $lib/lib$name.a -o -e $lib/lib$name.so && {
 			    SYSLIBS="$SYSLIBS -l$name"
