--- postfix-2.8.5/src/global/dict_ldap.c.orig	2011-10-27 16:06:24.139024496 -0700
+++ postfix-2.8.5/src/global/dict_ldap.c	2011-10-27 16:13:54.959021674 -0700
@@ -498,6 +498,7 @@
 static int dict_ldap_result(LDAP *ld, int msgid, int timeout, LDAPMessage **res)
 {
     struct timeval mytimeval;
+    int err;
 
     mytimeval.tv_sec = timeout;
     mytimeval.tv_usec = 0;
@@ -505,10 +506,12 @@
 #define GET_ALL 1
     if (ldap_result(ld, msgid, GET_ALL, &mytimeval, res) == -1)
 	return (dict_ldap_get_errno(ld));
-
-    if (dict_ldap_get_errno(ld) == LDAP_TIMEOUT) {
-	(void) dict_ldap_abandon(ld, msgid);
-	return (dict_ldap_set_errno(ld, LDAP_TIMEOUT));
+    if ((err = dict_ldap_get_errno(ld)) != LDAP_SUCCESS) {
+      if (err == LDAP_TIMEOUT) {
+        (void) dict_ldap_abandon(ld, msgid);
+        return (dict_ldap_set_errno(ld, LDAP_TIMEOUT));
+      }
+      return err;
     }
     return LDAP_SUCCESS;
 }
@@ -566,8 +569,11 @@
 			       &res)) != LDAP_SUCCESS)
 	return (rc);
 
-#define FREE_RESULT 1
-    return (ldap_parse_sasl_bind_result(dict_ldap->ld, res, 0, FREE_RESULT));
+#define FREE_RESULT 0
+    if ((rc = ldap_parse_sasl_bind_result(dict_ldap->ld, res, 0, FREE_RESULT)) != LDAP_SUCCESS)
+	return (rc);
+
+    return(ldap_result2error( dict_ldap->ld, res, 1 ));
 }
 
 /* search_st - Synchronous search with timeout */
