--- openldap-2.4.21/servers/slapd/back-bdb/dn2id.c.orig	2009/06/19 05:11:01	1.167
+++ openldap-2.4.21/servers/slapd/back-bdb/dn2id.c	2010/01/11 21:14:15	1.168
@@ -23,6 +23,7 @@
 #include "idl.h"
 #include "lutil.h"
 
+#if 0
 #define bdb_dn2id_lock					BDB_SYMBOL(dn2id_lock)
 
 static int
@@ -48,6 +49,9 @@ bdb_dn2id_lock( struct bdb_info *bdb, st
 					&lockobj, db_rw, lock);
 	return rc;
 }
+#else
+#define	bdb_dn2id_lock(a,b,c,d,e)	0
+#endif
 
 #ifndef BDB_HIER
 int
