--- openldap-2.4.33/libraries/liblmdb/mdb.c.orig	2013-02-06 10:53:58.087399407 -0800
+++ openldap-2.4.33/libraries/liblmdb/mdb.c	2013-02-06 10:54:02.871267137 -0800
@@ -1370,7 +1370,7 @@
 			MDB_oldpages *mop = txn->mt_env->me_pghead;
 			if (num > 1) {
 				MDB_cursor m2;
-				int retry = 500, readit = 0, n2 = num-1;
+				int retry = 20000, readit = 0, n2 = num-1;
 				unsigned int i, j, k;
 
 				/* If current list is too short, must fetch more and coalesce */
