--- openldap-2.4.23/servers/slapd/back-bdb/idl.h.orig	2011-02-17 16:32:02.598593211 -0800
+++ openldap-2.4.23/servers/slapd/back-bdb/idl.h	2011-02-17 16:32:08.937757993 -0800
@@ -20,7 +20,7 @@
 /* IDL sizes - likely should be even bigger
  *   limiting factors: sizeof(ID), thread stack size
  */
-#define	BDB_IDL_LOGN	16	/* DB_SIZE is 2^16, UM_SIZE is 2^17 */
+#define	BDB_IDL_LOGN	17	/* DB_SIZE is 2^16, UM_SIZE is 2^17 */
 #define BDB_IDL_DB_SIZE		(1<<BDB_IDL_LOGN)
 #define BDB_IDL_UM_SIZE		(1<<(BDB_IDL_LOGN+1))
 #define BDB_IDL_UM_SIZEOF	(BDB_IDL_UM_SIZE * sizeof(ID))
--- openldap-2.4.29/servers/slapd/back-mdb/idl.h.orig	2012-02-23 12:46:08.375760521 -0800
+++ openldap-2.4.29/servers/slapd/back-mdb/idl.h	2012-02-23 12:46:15.835762339 -0800
@@ -20,7 +20,7 @@
 /* IDL sizes - likely should be even bigger
  *   limiting factors: sizeof(ID), thread stack size
  */
-#define	MDB_IDL_LOGN	16	/* DB_SIZE is 2^16, UM_SIZE is 2^17 */
+#define	MDB_IDL_LOGN	17	/* DB_SIZE is 2^16, UM_SIZE is 2^17 */
 #define MDB_IDL_DB_SIZE		(1<<MDB_IDL_LOGN)
 #define MDB_IDL_UM_SIZE		(1<<(MDB_IDL_LOGN+1))
 #define MDB_IDL_UM_SIZEOF	(MDB_IDL_UM_SIZE * sizeof(ID))
--- openldap-2.4.31.2z/include/ldap_pvt_thread.h.orig	2012-05-08 12:39:43.448475644 -0700
+++ openldap-2.4.31.2z/include/ldap_pvt_thread.h	2012-05-08 12:40:09.878482861 -0700
@@ -68,7 +68,7 @@
 #	undef LDAP_PVT_THREAD_SET_STACK_SIZE
 #elif !defined( LDAP_PVT_THREAD_STACK_SIZE )
 	/* LARGE stack. Will be twice as large on 64 bit machine. */
-#	define LDAP_PVT_THREAD_STACK_SIZE ( 1 * 1024 * 1024 * sizeof(void *) )
+#	define LDAP_PVT_THREAD_STACK_SIZE ( 2 * 1024 * 1024 * sizeof(void *) )
 #endif
 #endif /* !LDAP_PVT_THREAD_H_DONE */
 
--- openldap-2.4.31/libraries/libmdb/midl.h.orig	2012-05-08 12:51:07.558476745 -0700
+++ openldap-2.4.31/libraries/libmdb/midl.h	2012-05-08 12:51:20.158477565 -0700
@@ -53,7 +53,7 @@
 /* IDL sizes - likely should be even bigger
  *   limiting factors: sizeof(ID), thread stack size
  */
-#define	MDB_IDL_LOGN	16	/* DB_SIZE is 2^16, UM_SIZE is 2^17 */
+#define	MDB_IDL_LOGN	17	/* DB_SIZE is 2^16, UM_SIZE is 2^17 */
 #define MDB_IDL_DB_SIZE		(1<<MDB_IDL_LOGN)
 #define MDB_IDL_UM_SIZE		(1<<(MDB_IDL_LOGN+1))
 #define MDB_IDL_UM_SIZEOF	(MDB_IDL_UM_SIZE * sizeof(ID))
--- openldap-2.4.31/servers/slapd/back-mdb/delete.c.orig	2012-05-08 16:49:15.478476029 -0700
+++ openldap-2.4.31/servers/slapd/back-mdb/delete.c	2012-05-08 16:49:36.658475272 -0700
@@ -419,7 +419,7 @@
 	}
 
 	if( rs->sr_err != 0 ) {
-		Debug( LDAP_DEBUG_TRACE,
+		Debug( LDAP_DEBUG_ANY,
 			LDAP_XSTRING(mdb_delete) ": txn_%s failed: %s (%d)\n",
 			op->o_noop ? "abort (no-op)" : "commit",
 			mdb_strerror(rs->sr_err), rs->sr_err );
