--- openldap-2.3.40/servers/slapd/syncrepl.c.orig	2007-12-10 22:08:43.000000000 -0800
+++ openldap-2.3.40/servers/slapd/syncrepl.c	2008-02-12 19:52:50.000000000 -0800
@@ -724,7 +724,11 @@
 					if ( syncrepl_message_to_op( si, op, msg ) == LDAP_SUCCESS &&
 						!BER_BVISNULL( &syncCookie.ctxcsn ) ) {
 						syncrepl_updateCookie( si, op, psub, &syncCookie );
+					} else if ( rc == LDAP_NO_SUCH_OBJECT ) {
+						rc = LDAP_SYNC_REFRESH_REQUIRED;
+						si->si_logstate = SYNCLOG_FALLBACK;
 					}
+						
 				} else if ( syncrepl_message_to_entry( si, op, msg,
 					&modlist, &entry, syncstate ) == LDAP_SUCCESS ) {
 					rc_efree = syncrepl_entry( si, op, entry, &modlist,
