--- openldap-2.3.42/servers/slapd/daemon.c.orig	2008-07-07 10:16:59.000000000 -0700
+++ openldap-2.3.42/servers/slapd/daemon.c	2008-07-07 10:19:52.000000000 -0700
@@ -810,9 +810,9 @@
 {
 	ldap_pvt_thread_mutex_lock( &slap_daemon.sd_mutex );
 
-	assert( SLAP_SOCK_IS_ACTIVE( s ));
-
 	if ( SLAP_SOCK_IS_WRITE( s )) {
+		assert( SLAP_SOCK_IS_ACTIVE( s ));
+
 		SLAP_SOCK_CLR_WRITE( s );
 		slap_daemon.sd_nwriters--;
 	}
