--- openldap-2.3.39/servers/slapd/back-bdb/tools.c.orig	2007-11-28 14:18:44.000000000 -0800
+++ openldap-2.3.39/servers/slapd/back-bdb/tools.c	2007-11-28 14:19:46.000000000 -0800
@@ -664,6 +664,10 @@
 		(long) e->e_id, e->e_dn, 0 );
 
 	if (! (slapMode & SLAP_TOOL_QUICK)) {
+		if( cursor ) {
+			cursor->c_close( cursor );
+			cursor = NULL;
+		}
 	rc = TXN_BEGIN( bdb->bi_dbenv, NULL, &tid, 
 		bdb->bi_db_opflags );
 	if( rc != 0 ) {
