--- openldap-2.4.16/include/lber.h.orig	2009-05-01 12:32:32.000000000 -0700
+++ openldap-2.4.16/include/lber.h	2009-05-01 12:32:38.000000000 -0700
@@ -629,6 +629,14 @@
 ber_strdup LDAP_P((
 	LDAP_CONST char * ));
 
+LBER_F( ber_len_t )
+ber_strnlen LDAP_P((
+	LDAP_CONST char *s, ber_len_t len ));
+
+LBER_F( char * )
+ber_strndup LDAP_P((
+	LDAP_CONST char *s, ber_len_t l ));
+
 LBER_F( struct berval * )
 ber_bvreplace LDAP_P((
 	struct berval *dst, LDAP_CONST struct berval *src ));
--- openldap-2.4.16/libraries/liblber/memory.c.orig	2009-05-01 12:35:06.000000000 -0700
+++ openldap-2.4.16/libraries/liblber/memory.c	2009-05-01 12:35:14.000000000 -0700
@@ -651,6 +651,16 @@
 	return ber_strdup_x( s, NULL );
 }
 
+ber_len_t
+ber_strnlen( LDAP_CONST char *s, ber_len_t len )
+{
+	ber_len_t l;
+
+	for ( l = 0; l < len && s[l] != '\0'; l++ ) ;
+
+	return l;
+}
+
 char *
 ber_strndup_x( LDAP_CONST char *s, ber_len_t l, void *ctx )
 {
@@ -666,11 +676,7 @@
 		return NULL;
 	}
 
-	len = strlen( s );
-
-	if ( len > l ) {
-		len = l;
-	}
+	len = ber_strnlen( s, l );
 
 	if ( (p = ber_memalloc_x( len + 1, ctx )) == NULL ) {
 		ber_errno = LBER_ERROR_MEMORY;
