--- openldap-2.4.21/servers/slapd/config.c.orig	2009/12/18 18:07:16	1.512
+++ openldap-2.4.21/servers/slapd/config.c	2010/02/02 13:43:21	1.513
@@ -1545,23 +1545,31 @@ slap_cf_aux_table_unparse( void *src, st
 			break;
 
 		case 'x':
-			*ptr++ = ' ';
-			ptr = lutil_strcopy( ptr, tab->key.bv_val );
-			if ( tab->quote ) *ptr++ = '"';
-			if ( tab->aux != NULL ) {
-				struct berval value;
-				slap_cf_aux_table_parse_x *func = (slap_cf_aux_table_parse_x *)tab->aux;
-				int rc;
+			{
+				char *saveptr=ptr;
+				*ptr++ = ' ';
+				ptr = lutil_strcopy( ptr, tab->key.bv_val );
+				if ( tab->quote ) *ptr++ = '"';
+				if ( tab->aux != NULL ) {
+					struct berval value;
+					slap_cf_aux_table_parse_x *func = (slap_cf_aux_table_parse_x *)tab->aux;
+					int rc;
 
-				value.bv_val = ptr;
-				value.bv_len = buf + sizeof( buf ) - ptr;
+					value.bv_val = ptr;
+					value.bv_len = buf + sizeof( buf ) - ptr;
 
-				rc = func( &value, (void *)((char *)src + tab->off), tab, "(unparse)", 1 );
-				if ( rc == 0 ) {
-					ptr += value.bv_len;
+					rc = func( &value, (void *)((char *)src + tab->off), tab, "(unparse)", 1 );
+					if ( rc == 0 ) {
+						if (value.bv_len) {
+							ptr += value.bv_len;
+						} else {
+							ptr = saveptr;
+							break;
+						}
+					}
 				}
+				if ( tab->quote ) *ptr++ = '"';
 			}
-			if ( tab->quote ) *ptr++ = '"';
 			break;
 
 		default:
