--- openldap-2.4.16/servers/slapd/back-bdb/cache.c	19 Mar 2009 05:06:37 -0000	1.199
+++ openldap-2.4.16/servers/slapd/back-bdb/cache.c	1 May 2009 02:36:39 -0000	1.200
@@ -1345,6 +1345,9 @@ bdb_cache_delete_cleanup(
 {
 	/* Enter with ei locked */
 
+	/* already freed? */
+	if ( !ei->bei_parent ) return;
+
 	if ( ei->bei_e ) {
 		ei->bei_e->e_private = NULL;
 #ifdef SLAP_ZONE_ALLOC
@@ -1368,6 +1371,10 @@ bdb_cache_delete_internal(
 	int rc = 0;	/* return code */
 	int decr_leaf = 0;
 
+	/* already freed? */
+	if ( !e->bei_parent )
+		return -1;
+
 	/* Lock the parent's kids tree */
 	bdb_cache_entryinfo_lock( e->bei_parent );
 
