--- openldap-2.4.21/servers/slapd/syncprov.c.orig	5 Jan 2010 02:50:10 -0000	1.308
+++ openldap-2.4.21/servers/slapd/syncprov.c	21 Jan 2010 12:33:21 -0000	1.309
@@ -946,11 +946,11 @@ syncprov_qplay( Operation *op, syncops *
 
 			rc = syncprov_sendresp( op, &opc, so, sr->s_mode );
 
-			if ( opc.se ) {
-				if ( !dec_mutexint( opc.se->e_private )) {
-					opc.se->e_private = NULL;
-					entry_free ( opc.se );
-				}
+		}
+		if ( sr->s_e ) {
+			if ( !dec_mutexint( sr->s_e->e_private )) {
+				sr->s_e->e_private = NULL;
+				entry_free ( sr->s_e );
 			}
 		}
 
