--- openldap-2.4.26/servers/slapd/back-bdb/dn2id.c.orig	2011-08-25 15:12:57.012098428 -0700
+++ openldap-2.4.26/servers/slapd/back-bdb/dn2id.c	2011-08-25 15:13:47.242098291 -0700
@@ -945,6 +945,7 @@
 			cx->rc = bdb_idl_cache_get(cx->bdb, cx->db, &cx->key, ids);
 			if ( cx->rc == LDAP_SUCCESS ) {
 				if ( cx->depth ) {
+					bdb_idl_delete( cx->tmp, cx->id );	/* ITS#6983, drop our own ID */
 					bdb_idl_append( cx->ids, cx->tmp );
 					cx->need_sort = 1;
 				}
--- openldap-2.4.26/servers/slapd/back-bdb/idl.c.orig	2011-08-25 15:13:58.212099794 -0700
+++ openldap-2.4.26/servers/slapd/back-bdb/idl.c	2011-08-25 15:14:50.632098843 -0700
@@ -219,7 +219,7 @@
 	return 0;
 }
 
-static int bdb_idl_delete( ID *ids, ID id )
+int bdb_idl_delete( ID *ids, ID id )
 {
 	unsigned x;
 
