From 970e7f8199fff70b313b1a25c33603f016830419 Mon Sep 17 00:00:00 2001
From: Howard Chu <hyc@openldap.org>
Date: Thu, 11 Oct 2012 16:45:42 -0700
Subject: [PATCH] ITS#7416 make sure to reopen attr DBs after env reopen

---
 servers/slapd/back-mdb/attr.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/servers/slapd/back-mdb/attr.c b/servers/slapd/back-mdb/attr.c
index 8faf17a..31683d2 100644
--- a/servers/slapd/back-mdb/attr.c
+++ b/servers/slapd/back-mdb/attr.c
@@ -162,8 +162,10 @@ mdb_attr_dbs_close(
 {
 	int i;
 	for ( i=0; i<mdb->mi_nattrs; i++ )
-		if ( mdb->mi_attrs[i]->ai_dbi )
+		if ( mdb->mi_attrs[i]->ai_dbi ) {
 			mdb_close( mdb->mi_dbenv, mdb->mi_attrs[i]->ai_dbi );
+			mdb->mi_attrs[i]->ai_dbi = 0;
+		}
 }
 
 int
-- 
1.7.4.2

