From 11cf13bf9d50259e9990be9ee08506b151e2a907 Mon Sep 17 00:00:00 2001
From: Howard Chu <hyc@openldap.org>
Date: Fri, 17 Aug 2012 13:38:47 -0700
Subject: [PATCH] Fix slapadd -q -w double-free

---
 servers/slapd/back-mdb/tools.c |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/servers/slapd/back-mdb/tools.c b/servers/slapd/back-mdb/tools.c
index ec62c50..bc2f401 100644
--- a/servers/slapd/back-mdb/tools.c
+++ b/servers/slapd/back-mdb/tools.c
@@ -960,6 +960,7 @@ done:
 		e->e_id = NOID;
 	}
 	txn = NULL;
+	idcursor = NULL;
 
 	return e->e_id;
 }
-- 
1.7.4.2

