--- openldap-2.4.28/servers/slapd/bconfig.c.orig	2011-12-05 13:44:36.350769813 -0800
+++ openldap-2.4.28/servers/slapd/bconfig.c	2011-12-05 13:46:46.150768067 -0800
@@ -1661,7 +1661,7 @@
 				/* else prev is NULL, append to end of global list */
 			}
 			if(parse_oc(c, &oc, prev)) return(1);
-			if (!cfn->c_oc_head) cfn->c_oc_head = oc;
+			if (!cfn->c_oc_head || !c->valx) cfn->c_oc_head = oc;
 			if (cfn->c_oc_tail == prev) cfn->c_oc_tail = oc;
 			}
 			break;
@@ -1694,7 +1694,7 @@
 				/* else prev is NULL, append to end of global list */
 			}
 			if(parse_at(c, &at, prev)) return(1);
-			if (!cfn->c_at_head) cfn->c_at_head = at;
+			if (!cfn->c_at_head || !c->valx) cfn->c_at_head = at;
 			if (cfn->c_at_tail == prev) cfn->c_at_tail = at;
 			}
 			break;
@@ -1727,7 +1727,7 @@
 				/* else prev is NULL, append to end of global list */
 			}
 			if ( parse_syn( c, &syn, prev ) ) return(1);
-			if ( !cfn->c_syn_head ) cfn->c_syn_head = syn;
+			if ( !cfn->c_syn_head || !c->valx ) cfn->c_syn_head = syn;
 			if ( cfn->c_syn_tail == prev ) cfn->c_syn_tail = syn;
 			}
 			break;
