--- openldap-2.4.31/libraries/libmdb/mdb.c.orig	2012-04-03 12:24:39.871602529 -0700
+++ openldap-2.4.31/libraries/libmdb/mdb.c	2012-04-03 12:27:40.101605011 -0700
@@ -5054,11 +5054,11 @@
 static int
 mdb_update_key(MDB_page *mp, indx_t indx, MDB_val *key)
 {
-	indx_t			 ptr, i, numkeys;
-	int			 delta;
-	size_t			 len;
 	MDB_node		*node;
 	char			*base;
+	size_t			 len;
+	int			 delta;
+	indx_t			 ptr, i, numkeys;
 	DKBUF;
 
 	node = NODEPTR(mp, indx);
@@ -5078,6 +5078,7 @@
 #endif
 
 	delta = key->mv_size - node->mn_ksize;
+	delta += (delta & 1);
 	if (delta) {
 		if (delta > 0 && SIZELEFT(mp) < delta) {
 			DPRINTF("OUCH! Not enough room, delta = %d", delta);
