--- openldap-2.4.34/servers/slapd/back-mdb/idl.c.orig	2013-03-18 12:18:21.411273449 -0700
+++ openldap-2.4.34/servers/slapd/back-mdb/idl.c	2013-03-18 12:19:53.557094528 -0700
@@ -1074,7 +1074,7 @@
 			ids[l+1] = ids[j];
 			ids[j] = a;
 			jstack += 2;
-			if (ir-i+1 >= j-1) {
+			if (ir-i+1 >= j-l) {
 				istack[jstack] = ir;
 				istack[jstack-1] = i;
 				ir = j-1;
--- openldap-2.4.34/servers/slapd/modify.c.orig	2013-03-18 12:20:05.777104059 -0700
+++ openldap-2.4.34/servers/slapd/modify.c	2013-03-18 12:20:23.897093194 -0700
@@ -800,7 +800,7 @@
 				break;
 			EXCH(l+1,j);
 			jstack += 2;
-			if (ir-i+1 >= j) {
+			if (ir-i+1 > j-l) {
 				istack[jstack] = ir;
 				istack[jstack-1] = i;
 				ir = j;
