--- openldap-2.4.28/servers/slapd/connection.c.orig	2011-11-25 10:52:29.000000000 -0800
+++ openldap-2.4.28/servers/slapd/connection.c	2012-01-23 11:56:41.360769365 -0800
@@ -225,7 +225,6 @@
 		 */
 		if(( c->c_n_ops_executing && !c->c_writewaiter)
 			|| c->c_conn_state == SLAP_C_CLIENT ) {
-			connection_done( c );
 			continue;
 		}
 
@@ -247,8 +246,8 @@
 				continue;
 			}
 		}
-		connection_done( c );
 	}
+	connection_done( c );
 	if ( old && !writers )
 		slapd_clr_writetime( old );
 
@@ -270,12 +269,12 @@
 		 */
 		if(( c->c_n_ops_executing && !c->c_writewaiter)
 			|| c->c_conn_state == SLAP_C_CLIENT ) {
-			connection_done( c );
 			continue;
 		}
 		connection_closing( c, "dropping" );
 		connection_close( c );
 	}
+	connection_done( c );
 }
 
 static Connection* connection_get( ber_socket_t s )
@@ -1893,8 +1892,6 @@
 
 	assert( connections != NULL );
 
-	slapd_clr_write( s, 0 );
-
 	c = connection_get( s );
 	if( c == NULL ) {
 		Debug( LDAP_DEBUG_ANY,
@@ -1903,6 +1900,8 @@
 		return -1;
 	}
 
+	slapd_clr_write( s, 0 );
+
 #ifdef HAVE_TLS
 	if ( c->c_is_tls && c->c_needs_tls_accept ) {
 		connection_return( c );
