From 784757faa91b4570465e7fc6453305324e686a0a Mon Sep 17 00:00:00 2001
From: Howard Chu <hyc@symas.com>
Date: Fri, 7 Dec 2012 12:06:30 -0800
Subject: [PATCH] ITS#7465 fail if both -a and -s specified

---
 libraries/liblmdb/mdb_stat.c |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/libraries/liblmdb/mdb_stat.c b/libraries/liblmdb/mdb_stat.c
index 4dfcf49..e9981f6 100644
--- a/libraries/liblmdb/mdb_stat.c
+++ b/libraries/liblmdb/mdb_stat.c
@@ -62,6 +62,8 @@ int main(int argc, char *argv[])
 	while ((i = getopt(argc, argv, "aefns:")) != EOF) {
 		switch(i) {
 		case 'a':
+			if (subname)
+				usage(prog);
 			alldbs++;
 			break;
 		case 'e':
@@ -74,6 +76,8 @@ int main(int argc, char *argv[])
 			envflags |= MDB_NOSUBDIR;
 			break;
 		case 's':
+			if (alldbs)
+				usage(prog);
 			subname = optarg;
 			break;
 		default:
-- 
1.7.4.2

