--- openldap-2.4.28/tests/scripts/test001-slapadd.orig	2011-12-05 13:49:46.970767486 -0800
+++ openldap-2.4.28/tests/scripts/test001-slapadd	2011-12-05 13:51:06.130771744 -0800
@@ -71,6 +71,9 @@
 	exit 1
 fi
 
+if test $BACKEND = ldif ; then
+	echo "Skipping test of unordered slapadd (unsupported in ldif backend)"
+else
 kill -HUP $KILLPIDS
 
 rm -f $DBDIR1/*
@@ -110,10 +113,10 @@
 	sleep 5
 done
 
-test $KILLSERVERS != no && kill -HUP $KILLPIDS
 
 if test $RC != 0 ; then
 	echo "ldapsearch failed ($RC)!"
+	test $KILLSERVERS != no && kill -HUP $KILLPIDS
 	exit $RC
 fi
 
@@ -128,9 +131,14 @@
 	echo "comparison failed - database was not created correctly"
 	echo $SEARCHFLT $LDIFFLT
 	$DIFF $SEARCHFLT $LDIFFLT
+	test $KILLSERVERS != no && kill -HUP $KILLPIDS
 	exit 1
 fi
 
+fi
+
+test $KILLSERVERS != no && kill -HUP $KILLPIDS
+
 echo ">>>>> Test succeeded"
 
 test $KILLSERVERS != no && wait
