--- openldap-2.4.28/libraries/libmdb/mdb.c.orig	2011-12-21 11:47:45.321599439 -0800
+++ openldap-2.4.28/libraries/libmdb/mdb.c	2012-01-04 17:20:49.173284042 -0800
@@ -4342,7 +4342,7 @@
 	size_t		 sz;
 
 	sz = LEAFSIZE(key, data);
-	if (data->mv_size >= env->me_psize / MDB_MINKEYS) {
+	if (sz >= env->me_psize / MDB_MINKEYS) {
 		/* put on overflow page */
 		sz -= data->mv_size - sizeof(pgno_t);
 	}
@@ -5564,7 +5564,8 @@
 		/* Insert in right sibling. */
 		/* Reset insert index for right sibling. */
 			j = (i == newindx && ins_new);
-			mc->mc_pg[mc->mc_top] = rp;
+			if (i != newindx || ins_new)
+				mc->mc_pg[mc->mc_top] = rp;
 		}
 
 		if (i == newindx && !ins_new) {
