--- openldap-2.4.28/servers/slapd/operation.c.orig	2011-12-15 12:34:56.910769700 -0800
+++ openldap-2.4.28/servers/slapd/operation.c	2011-12-15 12:35:16.960765677 -0800
@@ -129,6 +129,7 @@
 	op->o_bd = NULL;
 	BER_BVZERO( &op->o_req_dn );
 	BER_BVZERO( &op->o_req_ndn );
+	memset( op->o_hdr, 0, sizeof( *op->o_hdr ));
 	memset( &op->o_request, 0, sizeof( op->o_request ));
 	memset( &op->o_do_not_cache, 0, sizeof( Operation ) - offsetof( Operation, o_do_not_cache ));
 	memset( opbuf->ob_controls, 0, sizeof( opbuf->ob_controls ));
