--- openldap-2.4.28/servers/slapd/back-mdb/id2entry.c.orig	2011-12-05 13:41:11.740768167 -0800
+++ openldap-2.4.28/servers/slapd/back-mdb/id2entry.c	2011-12-05 13:42:16.150775197 -0800
@@ -217,14 +217,14 @@
 )
 {
 	if ( e->e_private ) {
-		if ( slapMode & SLAP_TOOL_MODE ) {
-			ch_free( e->e_nname.bv_val );
-			ch_free( e->e_name.bv_val );
-			ch_free( e );
-		} else {
+		if ( op->o_hdr ) {
 			op->o_tmpfree( e->e_nname.bv_val, op->o_tmpmemctx );
 			op->o_tmpfree( e->e_name.bv_val, op->o_tmpmemctx );
 			op->o_tmpfree( e, op->o_tmpmemctx );
+		} else {
+			ch_free( e->e_nname.bv_val );
+			ch_free( e->e_name.bv_val );
+			ch_free( e );
 		}
 	} else {
 		entry_free( e );
--- openldap-2.4.28/servers/slapd/overlays/dds.c.orig	2011-12-05 13:43:24.450769362 -0800
+++ openldap-2.4.28/servers/slapd/overlays/dds.c	2011-12-05 13:43:48.210766480 -0800
@@ -1708,6 +1708,9 @@
 	int		rc = 0;
 	void		*thrctx = ldap_pvt_thread_pool_context();
 
+	if ( slapMode & SLAP_TOOL_MODE )
+		return 0;
+
 	if ( DDS_OFF( di ) ) {
 		goto done;
 	}
