--- openldap-2.4.39/libraries/liblmdb/lmdb.h.orig	2014-01-13 15:38:19.265630402 -0600
+++ openldap-2.4.39/libraries/liblmdb/lmdb.h	2014-01-13 15:37:51.045631342 -0600
@@ -184,7 +184,7 @@
 /** Library minor version */
 #define MDB_VERSION_MINOR	9
 /** Library patch version */
-#define MDB_VERSION_PATCH	10
+#define MDB_VERSION_PATCH	11
 
 /** Combine args a,b,c into a single integer for easy version comparisons */
 #define MDB_VERINT(a,b,c)	(((a) << 24) | ((b) << 16) | (c))
