--- openldap-2.4.33.2z/servers/slapd/back-mdb/init.c.orig	2012-09-18 11:41:28.594792966 -0700
+++ openldap-2.4.33.2z/servers/slapd/back-mdb/init.c	2012-09-18 11:42:02.794791070 -0700
@@ -55,7 +55,11 @@
 
 	/* DBEnv parameters */
 	mdb->mi_dbenv_home = ch_strdup( SLAPD_DEFAULT_DB_DIR );
-	mdb->mi_dbenv_flags = 0;
+	if ( slapMode & SLAP_TOOL_READONLY) {
+	  mdb->mi_dbenv_flags = 0;
+	} else {
+	  mdb->mi_dbenv_flags = MDB_WRITEMAP;
+	}
 	mdb->mi_dbenv_mode = SLAPD_DEFAULT_DB_MODE;
 
 	mdb->mi_search_stack_depth = DEFAULT_SEARCH_STACK_DEPTH;
