--- openldap-2.4.29/servers/slapd/back-mdb/tools.c.orig	2012-02-29 17:26:21.314094712 -0800
+++ openldap-2.4.29/servers/slapd/back-mdb/tools.c	2012-02-29 17:30:09.411596643 -0800
@@ -335,8 +335,10 @@
 		rc = mdb_id2name( &op, txn, &idcursor, id, &dn, &ndn );
 		if ( rc  ) {
 			rc = LDAP_OTHER;
-			mdb_entry_return( &op, e );
-			e = NULL;
+			if ( e ) {
+				mdb_entry_return( &op, e );
+				e = NULL;
+			}
 			goto done;
 		}
 		if ( tool_base != NULL ) {
