--- curl-7.19.4/configure.orig	2009-04-08 11:29:56.000000000 -0700
+++ curl-7.19.4/configure	2009-04-08 11:30:21.000000000 -0700
@@ -23435,7 +23435,7 @@
         PKGTEST="no"
     PREFIX_OPENSSL=$OPT_SSL
     LIB_OPENSSL="$PREFIX_OPENSSL/lib$libsuff"
-    if  "$PREFIX_OPENSSL" != "/usr"  ; then
+    if test "$PREFIX_OPENSSL" != "/usr"  ; then
       LDFLAGS="$LDFLAGS -L$LIB_OPENSSL"
       CPPFLAGS="$CPPFLAGS -I$PREFIX_OPENSSL/include"
     fi
