--- clamav-0.97.7/configure.orig	2013-03-25 09:15:59.908637752 -0700
+++ clamav-0.97.7/configure	2013-03-25 09:23:26.041636309 -0700
@@ -15436,7 +15436,7 @@
 fi
 
 
-
+if test "$gcc_check" = "yes"; then
 { $as_echo "$as_me:$LINENO: checking for gcc bug PR28045" >&5
 $as_echo_n "checking for gcc bug PR28045... " >&6; }
 if test "$cross_compiling" = yes; then
@@ -15511,7 +15511,6 @@
 fi
 
 
-
 { $as_echo "$as_me:$LINENO: checking for gcc bug PR37573" >&5
 $as_echo_n "checking for gcc bug PR37573... " >&6; }
 if test "$cross_compiling" = yes; then
@@ -15638,6 +15637,7 @@
 rm -rf conftest.dSYM
 rm -f core *.core core.conftest.* gmon.out bb.out conftest$ac_exeext conftest.$ac_objext conftest.$ac_ext
 fi
+fi
 
 
 
@@ -19876,6 +19876,7 @@
 	esac
     fi
 
+if test "$gcc_check" = "yes"; then
     if test "$HAVE_LIBBZ2" = "yes"; then
 	{ $as_echo "$as_me:$LINENO: checking for CVE-2010-0405" >&5
 $as_echo_n "checking for CVE-2010-0405... " >&6; }
@@ -20047,6 +20048,8 @@
 $as_echo "$as_me: WARNING: ****** bzip2 support disabled" >&2;}
     fi
 fi
+fi
+LIBCLAMAV_LIBS="$LIBCLAMAV_LIBS $LTLIBBZ2"
 
 # Check whether --enable-unrar was given.
 if test "${enable_unrar+set}" = set; then
