--- clamav-0.95.3/configure.orig	2009-11-05 15:35:35.000000000 -0800
+++ clamav-0.95.3/configure	2009-11-05 15:36:56.000000000 -0800
@@ -12849,7 +12849,7 @@
   conftest.$ac_objext conftest.beam conftest.$ac_ext
 fi
 
-
+if test "$gcc_check" = "yes"; then
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking for gcc bug PR28045" >&5
 $as_echo_n "checking for gcc bug PR28045... " >&6; }
 if test "$cross_compiling" = yes; then :
@@ -12891,7 +12891,6 @@
   conftest.$ac_objext conftest.beam conftest.$ac_ext
 fi
 
-
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking for gcc bug PR37573" >&5
 $as_echo_n "checking for gcc bug PR37573... " >&6; }
 if test "$cross_compiling" = yes; then :
@@ -12986,7 +12985,7 @@
 rm -f core *.core core.conftest.* gmon.out bb.out conftest$ac_exeext \
   conftest.$ac_objext conftest.beam conftest.$ac_ext
 fi
-
+fi
 
 { $as_echo "$as_me:${as_lineno-$LINENO}: checking for ld --version-script" >&5
 $as_echo_n "checking for ld --version-script... " >&6; }
