--- altermime-0.3.10/qpe.c.orig	2008-11-16 00:45:45.000000000 -0800
+++ altermime-0.3.10/qpe.c	2011-09-02 15:14:50.571068401 -0700
@@ -97,7 +97,7 @@
 				op+= strlen(paragraph);// +3; /** jump the output + =\r\n **/
 				out_remaining-= (strlen(paragraph)); // Was +3, updated to fix Outlook problems
 
-				QPD fprintf(stdout, "Soft break (%d + %d > 76 char) for '%s'\n", current_line_length, charout_size, paragraph);
+				QPD fprintf(stdout, "Soft break (%d + %d > 76 char) for '%s'\n", (int)current_line_length, (int)charout_size, paragraph);
 				
 				/** reinitialize the paragraph **/
 				paragraph[0] = '\0';
@@ -108,7 +108,7 @@
 			}
 
 			snprintf(pp, pp_remaining, "%s", charout);
-			QPD fprintf(stdout,"charout='%s', size=%d, pp_remain=%d result='%s'\n", charout, charout_size, pp_remaining, paragraph);
+			QPD fprintf(stdout,"charout='%s', size=%d, pp_remain=%d result='%s'\n", charout, (int)charout_size, (int)pp_remaining, paragraph);
 			pp += charout_size;
 			pp_remaining -= charout_size;
 			p++;
@@ -149,13 +149,13 @@
 	out_size = in_size *3;
 	in_buffer = malloc( sizeof(char) *in_size +1);
 	if (in_buffer == NULL) {
-		QPD fprintf(stdout,"Error allocating %d bytes for input buffer\n", in_size);
+		QPD fprintf(stdout,"Error allocating %d bytes for input buffer\n", (int)in_size);
 		return -1;
 	}
 
 	out_buffer = malloc( sizeof(char) *out_size *3 +1);
 	if (in_buffer == NULL) {
-		QPD fprintf(stdout,"Error allocating %d bytes for output buffer\n", out_size);
+		QPD fprintf(stdout,"Error allocating %d bytes for output buffer\n", (int)out_size);
 		return -1;
 	}
 
@@ -169,7 +169,7 @@
 	  ** we segfault ;)  **/
 	*(in_buffer +in_size) = '\0';
 
-	QPD fprintf(stdout,"file %s is loaded, size = %d\n", fname, in_size);
+	QPD fprintf(stdout,"file %s is loaded, size = %d\n", fname, (int)in_size);
 
 	qp_encode( out_buffer, out_size, in_buffer, in_size );
 
