--- rrdtool-1.2.29/configure.ac.orig	2008-12-15 14:05:14.000000000 -0800
+++ rrdtool-1.2.29/configure.ac	2009-03-11 16:00:45.000000000 -0700
@@ -552,6 +552,7 @@
 AC_ARG_VAR(PERLCCFLAGS, [[] CC flags for Perl modules])
 AC_ARG_VAR(PERLLD, [[same as PERLCC] Linker for Perl modules])
 AC_ARG_VAR(PERLLDFLAGS, [[] LD flags for Perl modules])
+AC_ARG_VAR(PERLLDDLFLAGS, [[] LDD flags for Perl modules])
 
 if test "x$PERL" = "xno" -o  x$enable_perl = xno; then
 	COMP_PERL=
@@ -615,6 +616,10 @@
    if test ! -z "$PERLLDFLAGS"; then
        PERL_MAKE_OPTIONS="$PERL_MAKE_OPTIONS LDFLAGS=$PERLLDFLAGS"
    fi
+
+   if test ! -z "$PERLLDDLFLAGS"; then
+       PERL_MAKE_OPTIONS="$PERL_MAKE_OPTIONS LDDLFLAGS=$PERLLDDLFLAGS"
+   fi
 fi
         
 AC_ARG_WITH(perl-options,
