--- cyrus-sasl-2.1.25/saslauthd/Makefile.in.orig	2011-10-19 16:30:10.611876845 -0700
+++ cyrus-sasl-2.1.25/saslauthd/Makefile.in	2011-10-19 16:30:21.686037434 -0700
@@ -119,7 +119,7 @@
 CPP = @CPP@
 CPPFLAGS = @CPPFLAGS@
 CYGPATH_W = @CYGPATH_W@
-DEFS = @DEFS@ -DSASLAUTHD_CONF_FILE_DEFAULT=\"@sysconfdir@/saslauthd.conf\" -I. -I$(srcdir) -I..
+DEFS = @DEFS@ -DSASLAUTHD_CONF_FILE_DEFAULT=\"/opt/zimbra/conf/saslauthd.conf\" -I. -I$(srcdir) -I..
 DEPDIR = @DEPDIR@
 ECHO_C = @ECHO_C@
 ECHO_N = @ECHO_N@
--- cyrus-sasl-2.1.25/saslauthd/Makefile.am.orig	2011-10-19 16:30:14.021871044 -0700
+++ cyrus-sasl-2.1.25/saslauthd/Makefile.am	2011-10-19 16:30:30.016037595 -0700
@@ -26,7 +26,7 @@
 EXTRA_DIST	= saslauthd.8 saslauthd.mdoc config include \
 		  getnameinfo.c getaddrinfo.c LDAP_SASLAUTHD
 INCLUDES	= -I$(top_srcdir)/include -I$(top_builddir)/include -I$(top_srcdir)/../include
-DEFS            = @DEFS@ -DSASLAUTHD_CONF_FILE_DEFAULT=\"@sysconfdir@/saslauthd.conf\" -I. -I$(srcdir) -I..
+DEFS            = @DEFS@ -DSASLAUTHD_CONF_FILE_DEFAULT=\"/opt/zimbra/conf/saslauthd.conf\" -I. -I$(srcdir) -I..
 
 
 dist-hook: saslauthd.8
