--- gnupg-2.0.20/tests/openpgp/Makefile.in.orig	2013-08-12 16:07:54.811060411 -0700
+++ gnupg-2.0.20/tests/openpgp/Makefile.in	2013-08-12 16:08:52.643059639 -0700
@@ -645,7 +645,7 @@
 # start before all programs are built.
 ./gpg_dearmor: $(required_pgms)
 	echo '#!/bin/sh' >./gpg_dearmor
-	echo "../../g10/gpg2 --no-options --no-greeting \
+	echo "../../g10/gpg2 --homedir . --no-options --no-greeting \
              --no-secmem-warning --batch --dearmor" >>./gpg_dearmor
 	chmod 755 ./gpg_dearmor
 
--- gnupg-2.0.20/tests/openpgp/Makefile.am.orig	2013-08-12 16:07:51.022059530 -0700
+++ gnupg-2.0.20/tests/openpgp/Makefile.am	2013-08-12 16:08:24.534058731 -0700
@@ -79,7 +79,7 @@
 # start before all programs are built.
 ./gpg_dearmor: $(required_pgms)
 	echo '#!/bin/sh' >./gpg_dearmor
-	echo "../../g10/gpg2 --no-options --no-greeting \
+	echo "../../g10/gpg2 --homedir . --no-options --no-greeting \
              --no-secmem-warning --batch --dearmor" >>./gpg_dearmor
 	chmod 755 ./gpg_dearmor
 
